embrace-android-delivery/src/test/kotlin/io/embrace/android/embracesdk/internal/delivery/storage/PayloadStorageServiceImplTest.kt

```diff
@@ -0,0 +1,116 @@
+    private lateinit var service: PayloadStorageService
+    private lateinit var internalErrorService: FakeInternalErrorService
+    private lateinit var outputDir: File
+
+    @Before
+    fun setUp() {
+        internalErrorService = FakeInternalErrorService()
+        outputDir = Files.createTempDirectory("output").toFile()
+        outputDir.deleteRecursively()
+        service = PayloadStorageServiceImpl(internalErrorService, lazy { outputDir })
+    }
+
+    @Test
+    fun `test store and load object`() {
+        val fileContents = "test"
+
+        // store file
+        service.store(metadata) {
+            it.write(fileContents.toByteArray())
+        }
+        val files = outputDir.listFiles()
+        val file = files?.single() ?: error("File not found")
+        assertEquals(fileContents, GZIPInputStream(file.inputStream()).bufferedReader().readText())
+
+        // read file
+        val observed =
+            GZIPInputStream(service.loadPayloadAsStream(metadata)).bufferedReader().readText()
+        assertEquals(fileContents, observed)
+
+        // delete file
+        service.delete(metadata)
+        assertEquals(0, outputDir.listFiles()?.size)
+        assertNull(service.loadPayloadAsStream(metadata))
+    }
+
+    @Test
+    fun `delete non existent file`() {
+        service.delete(metadata) // no exception thrown
+        assertNull(outputDir.listFiles())
+    }
+
+    @Test
+    fun `load non existent file`() {
+        assertNull(service.loadPayloadAsStream(metadata))
+        assertNotNull(internalErrorService.throwables.single())
+    }
+
+    @Test
+    fun `exception in serialization action when storing`() {
+        service.store(metadata) {
+            error("Whoops!")
+        }
+        assertNotNull(internalErrorService.throwables.single())
+    }
+
+    @Test
+    fun `test objects pruned past limit`() {
+        service = PayloadStorageServiceImpl(internalErrorService, lazy { outputDir }, 4)
+
+        // exceed storage limit
+        listOf(
+            Pair(0L, CRASH),
+            Pair(0L, SESSION),
+            Pair(0L, LOG),
+            Pair(0L, NETWORK),
+            Pair(1000L, CRASH),
+            Pair(1000L, SESSION),
+            Pair(1000L, LOG),
+            Pair(1000L, NETWORK)
+        ).forEach {
+            val metadata = StoredTelemetryMetadata(it.first, UUID, it.second)
+            service.store(metadata) {
+                it.write("test".toByteArray())
+            }
+        }
+
+        val files = outputDir.listFiles() ?: error("No files found")
+        val outputs = files.map { StoredTelemetryMetadata.fromFilename(it.name).getOrThrow() }
+            .sortedWith(storedTelemetryComparator)
+            .map { Pair(it.timestamp, it.envelopeType) }
+
+        val expected = listOf(
+            Pair(0L, CRASH),
+            Pair(1000L, CRASH),
+            Pair(0L, SESSION),
+            Pair(1000L, NETWORK)
```

Comment:
I'm not convinced this is the right behavior - we should probably prioritise the `Pair(1000L, SESSION)` right, even if it means dropping the in-memory data?

Reply:
Yeah. If we have 500 sessions cached, there's no point trying to delivery a network capture.

That said, if a cache gets clogged up like that, we should try to notify ourselves that this is happening. If it's a lack of network, fair enough, but be should be able to side channel internal errors so they don't get lost if something bloats the cache like that.